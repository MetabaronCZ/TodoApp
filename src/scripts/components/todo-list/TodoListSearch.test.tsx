import React from 'react';
import { t } from 'i18next';
import { Provider } from 'react-redux';

import { fireEvent, render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';

import { TodoListSearch } from 'components/todo-list/TodoListSearch';

import { client } from 'client';

import { mockStore } from 'test/store';
import { withMockedProviders } from 'test/component';

const getTodoListSearch = (): JSX.Element => {
  const store = mockStore();

  return withMockedProviders(
    <Provider store={store}>
      <TodoListSearch />
    </Provider>,
  );
};

describe('components/todo-list/TodoListSearch', () => {
  it('should render correctly', () => {
    const { container } = render(getTodoListSearch());
    expect(container).toMatchSnapshot();
  });

  it('should not search todos without query', () => {
    const mockedTodoFetch = vi.spyOn(client.todo, 'get');
    const tree = render(getTodoListSearch());

    const input = tree.container.querySelector('input');
    expect(input).toBeInTheDocument();
    expect(input?.value).toEqual('');

    const submitButton = tree.getByRole('button', {
      name: (value) => value.includes(t('todoList.search')),
    });
    expect(submitButton).toBeInTheDocument();

    // trigger search
    fireEvent.click(submitButton);
    expect(mockedTodoFetch).toBeCalledTimes(0);
  });

  it('should search todos by given query', () => {
    const mockedTodoFetch = vi.spyOn(client.todo, 'get');
    const tree = render(getTodoListSearch());

    const input = tree.container.querySelector('input');
    expect(input).toBeInTheDocument();
    expect(input?.value).toEqual('');

    if (input) {
      fireEvent.change(input, { target: { value: 'Changed' } });
      expect(input.value).toEqual('Changed');
    }

    const submitButton = tree.getByRole('button', {
      name: (value) => value.includes(t('todoList.search')),
    });
    expect(submitButton).toBeInTheDocument();

    // trigger search
    fireEvent.click(submitButton);
    expect(mockedTodoFetch).toBeCalledTimes(1);

    expect(mockedTodoFetch).lastCalledWith(
      expect.objectContaining({ query: 'Changed' }),
    );
  });

  it('should clear query', () => {
    const mockedTodoFetch = vi.spyOn(client.todo, 'get');
    const tree = render(getTodoListSearch());

    const input = tree.container.querySelector('input');
    expect(input).toBeInTheDocument();
    expect(input?.value).toEqual('');

    if (input) {
      fireEvent.change(input, { target: { value: 'Changed' } });
      expect(input.value).toEqual('Changed');
    }

    const clearButton = tree.getByRole('button', {
      name: (value) => value.includes(t('cancel')),
    });
    expect(clearButton).toBeInTheDocument();

    // trigger search clear
    fireEvent.click(clearButton);
    expect(mockedTodoFetch).toBeCalledTimes(1);

    expect(mockedTodoFetch).lastCalledWith(
      expect.objectContaining({ query: '' }),
    );
  });
});
