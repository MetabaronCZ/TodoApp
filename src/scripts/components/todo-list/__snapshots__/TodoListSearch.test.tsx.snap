// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/todo-list/TodoListSearch should render correctly 1`] = `
.c3 {
  display: block;
  width: 100%;
  border: none;
  outline: none;
  cursor: pointer;
}

.c4 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: auto;
  padding: 5px 10px;
  background: #f0f0f0;
}

.c4:hover:not(:disabled) {
  background-color: #e0e0e0;
}

.c4:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c4:disabled {
  color: #c0c0c0;
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  font-family: Open Sans,Arial,sans-serif;
  color: #222;
  font-size: 14px;
  line-height: 20px;
  display: block;
  width: 100%;
  padding: 5px 10px;
  background: #fff;
  border: none;
  outline: none;
}

.c2:hover:not(:disabled) {
  background-color: #fcfcfc;
}

.c2:focus {
  outline: 1px dotted #222;
  outline-offset: -1px;
}

.c2:disabled {
  color: #c0c0c0;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c0 > *:first-child {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

<div>
  <form
    class="c0"
  >
    <div
      class="c1"
    >
      <input
        class="c2"
        maxlength="100"
        placeholder="Hledaný výraz..."
        type="text"
        value=""
      />
    </div>
    <button
      class="c3 c4"
      type="button"
    >
      <span
        class="c5"
      >
        Hledat
      </span>
    </button>
    <button
      class="c3 c4"
      disabled=""
      type="button"
    >
      <span
        class="c5"
      >
        Zrušit
      </span>
    </button>
  </form>
</div>
`;
